File: deps/UNR_AD/A4_PowerLLEL/src/mod_calcRHS.f90
module mod_calcRHS
    use mod_type,       only: fp
    use mod_parameters, only: dt, nhalo
    use mod_mpi,        only: sz
    use mod_mesh,       only: dx_inv, dy_inv, dzf_inv
    !$ use omp_lib
#ifdef _CUDA
    use cudafor
#endif
    use, intrinsic :: iso_c_binding

    implicit none

    interface
        subroutine calc_rhs(dt, dx_inv, dy_inv, dzf_inv, nhalo, nhalo_rhs, sz, u, v, w, rhs) bind(C, name='calc_rhs')
            import
            real(C_DOUBLE), value :: dt
            real(C_DOUBLE), value :: dx_inv
            real(C_DOUBLE), value :: dy_inv
            real(C_DOUBLE), dimension(*), intent(in) :: dzf_inv
            integer(C_INT), dimension(*), intent(in) :: nhalo, nhalo_rhs
            integer(C_INT), dimension(*), intent(in) :: sz
            real(C_DOUBLE), dimension(*), intent(in) :: u, v, w
            real(C_DOUBLE), dimension(*), intent(out) :: rhs
        end subroutine calc_rhs
    end interface

contains
    subroutine calcRHS(u, v, w, rhs)
        implicit none
        real(fp), dimension(1-nhalo(1):,1-nhalo(3):,1-nhalo(5):), intent(in ) :: u, v, w
        real(fp), dimension(0:,0:,0:), intent(out) :: rhs
        !@cuf attributes(managed) :: u, v, w, rhs
        !@cuf integer :: istat

        real(fp) :: dtidxi, dtidyi, dti
        integer  :: i, j, k

        dti = 1.0_fp/dt
        dtidxi = dti * dx_inv
        dtidyi = dti * dy_inv

#ifdef _CUDA
        !$cuf kernel do(3) <<<*,*>>>
#else
        !$OMP PARALLEL DO SCHEDULE(STATIC) &
        !$OMP DEFAULT(SHARED) PRIVATE(i, j, k)
#endif
        do k = 1, sz(3)
        do j = 1, sz(2)
        do i = 1, sz(1)
            rhs(i,j,k) = (u(i,j,k)-u(i-1,j,k))*dtidxi + &
                         (v(i,j,k)-v(i,j-1,k))*dtidyi + &
                         (w(i,j,k)-w(i,j,k-1))*dti*dzf_inv(k)
        enddo
        enddo
        enddo
#ifndef _CUDA
        !$OMP END PARALLEL DO
#endif
        !@cuf istat = cudaDeviceSynchronize()

        return
    end subroutine calcRHS

end module mod_calcRHS

File: deps/UNR_AD/A4_PowerLLEL/src/src_c/mod_calc_rhs.c
#include "bind.h"
#include <omp.h>


void calc_rhs(double dt, double dx_inv, double dy_inv, double *dzf_inv,
              int *nhalo, int *nhalo_rhs, int *sz, double *u, double *v,
              double *w, double *rhs) {

  double dti = 1.0 / dt;
  double dtidxi = dti * dx_inv;
  double dtidyi = dti * dy_inv;

  BIND1(nhalo)
  BIND1(nhalo_rhs)
  BIND1(sz)
  BIND1_EXT(dzf_inv, 1 - I1(nhalo, 5))
  BIND3_EXT(u, I1(sz, 1) + I1(nhalo, 1) + I1(nhalo, 2),
            I1(sz, 2) + I1(nhalo, 3) + I1(nhalo, 4), 1 - I1(nhalo, 1),
            1 - I1(nhalo, 3), 1 - I1(nhalo, 5))
  BIND3_EXT(v, I1(sz, 1) + I1(nhalo, 1) + I1(nhalo, 2),
            I1(sz, 2) + I1(nhalo, 3) + I1(nhalo, 4), 1 - I1(nhalo, 1),
            1 - I1(nhalo, 3), 1 - I1(nhalo, 5))
  BIND3_EXT(w, I1(sz, 1) + I1(nhalo, 1) + I1(nhalo, 2),
            I1(sz, 2) + I1(nhalo, 3) + I1(nhalo, 4), 1 - I1(nhalo, 1),
            1 - I1(nhalo, 3), 1 - I1(nhalo, 5))
  BIND3_EXT(rhs, I1(sz, 1) + I1(nhalo_rhs, 1) + I1(nhalo_rhs, 2),
            I1(sz, 2) + I1(nhalo_rhs, 3) + I1(nhalo_rhs, 4),
            1 - I1(nhalo_rhs, 1), 1 - I1(nhalo_rhs, 3), 1 - I1(nhalo_rhs, 5))

  const int sz0 = sz[0], sz1 = sz[1], sz2 = sz[2];
#ifdef USE_OMP_OFFLOAD
  #pragma omp target teams distribute parallel for collapse(3)
#else
  #pragma omp parallel for
#endif
  DO(k, 1, sz2) {
    DO(j, 1, sz1) {
      DO(i, 1, sz0) {
        I3(rhs, i, j, k) =
          (I3(u, i, j, k) - I3(u, i - 1, j, k)) * dtidxi +
          (I3(v, i, j, k) - I3(v, i, j - 1, k)) * dtidyi +
          (I3(w, i, j, k) - I3(w, i, j, k - 1)) * dti * I1(dzf_inv, k);
      }
    }
  }
}